app/services/order_purchaser.rb

```diff
@@ -0,0 +1,76 @@
+class OrderPurchaser
+
+  attr_reader :acting_as, :order, :order_in_past, :params, :user
+  alias acting_as? acting_as
+  alias order_in_past? order_in_past
+
+  delegate :order_notification_recipient, to: :order
+  delegate :quantities_changed?, to: :order_detail_updater
+
+  def initialize(acting_as:, order:, order_in_past:, params:, user:)
+    @acting_as = acting_as
+    @order = order
+    @order_in_past = order_in_past
+    @params = params
+    @user = user
+  end
+
+  def purchase!
+    order.transaction do
+      return if order_detail_updater.update! && quantities_changed?
+
+      validate_and_purchase!
+      order_in_the_past! if order_in_past?
+
+      Notifier.delay.order_receipt(user: order.user, order: order) if send_receipt?
+      Notifier.delay.order_notification(order, order_notification_recipient) if send_notification?
+    end
+  end
+
+  private
+
+  def order_detail_updater
+    @order_detail_updater ||= OrderDetailUpdater.new(order, order_update_params)
+  end
+
+  def order_in_the_past!
+    unless order.can_backdate_order_details?
+      raise NUCore::PurchaseException.new(I18n.t("controllers.orders.purchase.future_dating_error"))
+    end
+
+    # update order detail statuses if you've changed it while acting as
+    if acting_as? && order_status_id.present?
+      order.backdate_order_details!(user, order_status_id)
+    else
+      order.complete_past_reservations!
+    end
+  end
+
+  def order_status_id
+    params[:order_status_id]
+  end
+
+  def order_update_params
+    @order_update_params ||= OrderDetailUpdateParamHashExtractor.new(params).to_h
+  end
+
+  def quantities
+    order.order_details.order("order_details.id").pluck(:quantity)
+  end
+
+  def send_notification?
+    order_notification_recipient.present?
```

Comment:
We should probably not be sending the notification when placing an order on behalf of.